src/Compilers/Core/Portable/Compilation/Compilation.cs

```diff
@@ -3733,7 +3731,17 @@ public ImmutableArray<AssemblyIdentity> GetUnreferencedAssemblyIdentities(Diagno
                 return ImmutableArray<AssemblyIdentity>.Empty;
             }
 
-            var builder = ArrayBuilder<AssemblyIdentity>.GetInstance();
+            // Pre-calculate size to allow this code to only require a single ImmutableArray allocation.
+            var builderSize = 0;
+            foreach (var argument in diagnostic.Arguments)
+            {
+                if (argument is AssemblyIdentity id)
+                {
+                    builderSize++;
+                }
+            }
+
+            var builder = ImmutableArray.CreateBuilder<AssemblyIdentity>(builderSize);
```

Comment:
📝 This is relatively rare code, and when hit generally allocates less than 4 items. I would remove the `builderSize` calculation here and use `TemporaryArray<AssemblyIdentity>` instead.